### ISSUE TITLE
Error when talking to Tessel in DFU mode

### ISSUE THREAD

**Issue by tcr** [tessel/t2-cli#965]
We get an error when running Tessel commands with a Tessel in DFU mode. See: https://github.com/tessel/t2-cli/issues/827

I'd like to see the following:

* When running `t2 update`, we should say "To update or flash your Tessel in DFU mode, you probably want to use `t2 restore`."
* In any other command, warn `Detected a Tessel in DFU mode. Unplug and plug it in to boot normally.`

This is based off of not knowing there was a `t2 restore` command and the huge number of vintage Tessels I was upgrading by hand. 🙃

**Comment by tcr** [comment]
Actually I guess I'm confused about `t2 restore` in general:

* It only flashes OpenWRT, not OpenWRT and firmware. But it flashes it through the SAMD21, so could it also flash the SAMD21?
* It can't detect a device in DFU mode.
* When running `t2 update` with a device in DFU mode, it doesn't detect or suggest options. But we use DFU mode to flash the SAMD21. i.e. there's no "only flash SAMD21" option from the command line that's trivial.

I would expect the following, probably:

* When not booted, `t2 update` says there's a T2 booting.
* When booted, `t2 update` flashes version on OpenWRT and SAMD21.
* When in DFU, `t2 update` says there's a T2 in DFU mode. If you want to flash it in this mode, use `t2 restore`.
* When not booted, `t2 restore` succeeds in flashing both OpenWRT and SAMD21. (Maybe there are options to just do OpenWRT OR SAMD21?)
* When booted, `t2 restore` succeeds in flashing both OpenWRT and SAMD21.
* When in DFU, `t2 restore` succeeds in flashing both OpenWRT and SAMD21.

Right now, when I plug in a T2 in DFU mode, I'm not even sure what I can do with it from the command line. We have the option to physically boot it into this mode, but no way for the user to do anything with it.

**Comment by johnnyman727** [comment]
I think the assumption made when building those two commands was that most users wouldn't know what DFU mode is and shouldn't have to.

Both `t2 restore` and `t2 update` work properly from a Tessel in a standard boot state (not DFU).

I think that they could be made to work better if they could deal with a Tessel that's in DFU mode without throwing any errors.

Second best would be a warning telling the user to take the device out of DFU mode and to try the command again.

**Comment by rwaldron** [comment]
> Second best would be a warning telling the user to take the device out of DFU mode and to try the command again.

I would prefer is this was the approach

**Comment by tcr** [comment]
@johnnyman727 Yeah, I can probably submit a PR to accomplish either of those.

Nonetheless, say I have a borked SAMD21. The bootloader still works—if I throw it into DFU mode. But we have no commands which allow you to flash it that way, but t2 update would, if we could skip the OpenWRT updating part. Is that reasonable to add? "Your T2 is in DFU mode, only updating firmware."

**Comment by tcr** [comment]
And also, should `t2 restore` reflash the SAMD21 firmware as well? Why not?

**Comment by rwaldron** [comment]
> Right now there is no use for holding down reset to get into DFU mode, since the user can't do anything meaningful with it from the CLI.

I don't agree with this reasoning. The CLI's capabilities don't have an obligation to the button on the board. The on board button's "hold for DFU mode"  functionality is _vital_ for local t2-firmware development.

> And also, should t2 restore reflash the SAMD21 firmware as well? Why not?

IIRC, there is no Tessel remote connection available at the end of the restore operation. This also prevented the automatic rebooting of the board at the end of `t2 restore`

**Comment by tcr** [comment]
> I don't agree with this reasoning. The CLI's capabilities don't have an obligation to the button on the board. The on board button's "hold for DFU mode" functionality is vital for local t2-firmware development.

Fair point. I'm just saying from a user standpoint this is what we currently provide:

* My OpenWRT partition is screwed up. — Run `t2 restore`.
* My Firmware on the SAMD21 is screwed up. — Manually download a firmware file from builds.tessel.io. Install `dfu-util`. Switch into DFU mode on boot, then run the dfu-util command.

It's less discoverable, unless I'm missing something? This is just my experience this morning figuring out how to bring some old T2's to life. This process doesn't have to be added to t2-cli, but we should document it somewhere; and I think `t2 restore-dfu` or something would be useful in a pinch.

> IIRC, there is no Tessel remote connection available at the end of the restore operation.

Can we switch the USB alt setting at the end of the flashing period to DFU mode?

**Comment by tcr** [comment]
Okay, so here's what I propose I try. A PR that does the following:

1. A PR that warns the user to take the device out of DFU mode and to try the command again.
1. A change to `t2 restore` that flashes factory firmware on the SAMD21, then flashes the OpenWRT firmware, without requiring user intervention.

If I can't accomplish number 2) then I'd concede it's not worth the cost of integrating into the CLI. ;)

**Comment by rwaldron** [comment]
> A change to t2 restore that flashes factory firmware on the SAMD21, then flashes the OpenWRT firmware, without requiring user intervention.

In `t2 update`, the SAMD21 is updated last because doing so automatically reboots the SAMD21, which means the Tessel connection is lost.

**Comment by tcr** [comment]
`t2 restore` uses a different mechanism for flashing, specifically a dedicated USB alt setting.

**Comment by johnnyman727** [comment]
> A PR that warns the user to take the device out of DFU mode and to try the command again.

:+1:

> A change to t2 restore that flashes factory firmware on the SAMD21, then flashes the OpenWRT firmware, without requiring user intervention inbetween / any potential for additional issues.

Yeah, I think this makes sense. We didn't implement it originally because we were just focusing on borked OpenWRT images. As a user, I think I would expect it to update the firmware as well. Maybe more importantly, we should try to make sure the firmware running on the SAMD21 is compatible with the version of OpenWRT which necessitates flashing the firmware as well. You should be able to switch it into dfu mode from flash mode without a problem.

Also, it would be nice to not have to use `dfu-util` anymore. :+1:
